src/server/coprocessor/executor/aggregation.rs

```diff
@@ -0,0 +1,278 @@
+            gks: vec![],
+            gk_aggrs: map![],
+            cursor: 0,
+            executed: false,
+            ctx: ctx,
+            cols: cols,
+            src: src,
+        })
+    }
+
+    fn get_group_key(&mut self, eval: &mut Evaluator) -> Result<Vec<u8>> {
+        if self.group_by.is_empty() {
+            return Ok(SINGLE_GROUP.to_vec());
+        }
+        let mut vals = Vec::with_capacity(self.group_by.len());
+        for expr in &self.group_by {
+            let v = box_try!(eval.eval(&self.ctx, expr));
+            vals.push(v);
+        }
+        let res = box_try!(datum::encode_value(&vals));
+        Ok(res)
+    }
+
+    fn aggregate(&mut self) -> Result<()> {
+        while let Some(row) = try!(self.src.next()) {
+            let mut eval = Evaluator::default();
+            try!(inflate_with_col(&mut eval, &self.ctx, &row.data, &self.cols, row.handle));
+            let gk = try!(self.get_group_key(&mut eval));
+            match self.gk_aggrs.entry(gk.clone()) {
+                Entry::Vacant(e) => {
+                    let mut aggrs = Vec::with_capacity(self.aggr_func.len());
+                    for expr in &self.aggr_func {
+                        let mut aggr = try!(aggregate::build_aggr_func(expr));
+                        let vals = box_try!(eval.batch_eval(&self.ctx, expr.get_children()));
+                        try!(aggr.update(&self.ctx, vals));
+                        aggrs.push(aggr);
+                    }
+                    self.gks.push(gk);
+                    e.insert(aggrs);
+                }
+                Entry::Occupied(e) => {
+                    let aggrs = e.into_mut();
+                    for (expr, aggr) in self.aggr_func.iter().zip(aggrs) {
+                        let vals = box_try!(eval.batch_eval(&self.ctx, expr.get_children()));
+                        box_try!(aggr.update(&self.ctx, vals));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+
+impl<'a> Executor for AggregationExecutor<'a> {
+    fn next(&mut self) -> Result<Option<Row>> {
+        if !self.executed {
+            try!(self.aggregate());
+            self.executed = true;
+            assert_eq!(self.gks.len(), self.gk_aggrs.len());
+        }
+
+        if self.cursor >= self.gks.len() {
+            return Ok(None);
+        }
+        // calc all aggr func
+        let mut aggr_cols = Vec::with_capacity(2 * self.aggr_func.len());
+        let gk = &self.gks[self.cursor];
+        let mut aggrs = self.gk_aggrs.remove(gk).unwrap();
+        for aggr in &mut aggrs {
+            try!(aggr.calc(&mut aggr_cols));
+        }
+        // construct row data
+        let value_size = gk.len() + approximate_size(&aggr_cols, false);
+        let mut value = Vec::with_capacity(value_size);
+        let mut meta = HashMap::with_capacity(1 + aggr_cols.len());
+        let (mut id, mut offset) = (0, 0);
+        /// push gk col
+        value.extend_from_slice(gk);
```

Comment:
TiDB updates this format at this [PR](https://github.com/pingcap/tidb/pull/3483/files).